back/routes/adventure.js

```diff
@@ -73,51 +87,86 @@ async function getMean(experienceId, feedbackId) {
+        score += ( meanReviewScore / feedbackMaxValue.toJSON().maxValue ) * ( userPreference.activePreferences.length - indexPreference )
+      }
       if (userPreference.hasMinimumValue[indexPreference] && userPreference.minValues[indexPreference] > meanReviewScore) {
         passesFilter = false;
         break;
-      } else if (priority) {
-        const feedbackMaxValue = await new Parse.Query('Preference').equalTo('objectId', preferenceId).select('maxValue').first();
-        score += ( meanReviewScore / feedbackMaxValue.toJSON().maxValue ) * ( userPreference.activePreferences.length - indexPreference )
       }
-      indexPreference++;
     }
-    if (passesFilter && priority) restaurants.push({...restaurant, matchScore: score});
-    else if (passesFilter) restaurants.push(restaurant);
+    if (passesFilter) {
+      restaurants.push({
+        ...restaurant,
+        matchScore: priority ? score : undefined,
+      })
+    }
   }
   if (priority) {
     restaurants.sort((a, b) => {
       return b.matchScore - a.matchScore;
     })
   }
+  return restaurants;
+}
+
+// ----- Get filtered restaurants ------
+router.post('/restaurants', async (req, res) => {
+  const userPreference = await preferenceQuery(req.body.username);
+
+  // Contains the information of the distance from the db
+  const distanceFeedback = (await new Parse.Query('Preference').equalTo('displayText', "Distance").first()).toJSON();
+  
+  // Change meters to km in case distance is not defined by user
+  let distance = distanceFeedback.defaultValue * 1000;
+  // Get distance value by the user or by default in the database
+  if (userPreference.activePreferences.includes(distanceFeedback.objectId)) {
+    let index = userPreference.activePreferences.indexOf(distanceFeedback.objectId)
+    // Match km from UI options to meters in Google's API request
+    if (userPreference.hasMinimumValue[index]) distance = userPreference.minValues[index] * 1000;
+    // Delete distance from userPreferences (handled by Google Maps and by our database)
+    // The database records minValue even if it is not active (so if user turns it on, it goes back to that)
+    (userPreference.activePreferences).splice(index, 1);
+    (userPreference.hasMinimumValue).splice(index, 1);
+    (userPreference.minValues).splice(index, 1);
+  } 
+
+  // Get restaurants from Google (limiting distance)
+  const googleRestaurants = await getGoogleRestaurants(
+    distance,
+    req.body.lat,
+    req.body.lng,
+  );
+  // Get restaurants from our records (limiting distance)
+  let databaseRestaurants = await getDatabaseRestaurants(distance, req.body.lat, req.body.lng);
+  databaseRestaurants = databaseRestaurants.map((restaurant) => formatRestaurantForGoogleAPI(restaurant));
+
+  const allRestaurants = googleRestaurants.concat(databaseRestaurants)
+
+  // Calculate final restaurants
+  const restaurants = await filterAndRank(userPreference, allRestaurants);
   res.send(restaurants).status(200);
```

Comment:
nit (performance): Although this isn't in a loop, a similar comment to above. Do all of these calls actually depend on each other? Or could some of them be awaited in parallel, rather than one after the other?

Reply:
Thanks for all this! I included this feedback in an issue to improve later on!